### PR TITLE
Commit Availability Service

### PR DESCRIPTION
**Pull request by andrewchoi5**
**Commit Availability Service.**
1 - Addresses: Broken ConsumeService on latest release #195 https://github.com/linkedin/kafka-monitor/issues/195
2 - Addresses: https://github.com/linkedin/kafka-monitor/issues/204 
3 - Performs Consumer Offset Commit Availability using Metric Sensors. Does periodically, every few x seconds.
4 - In doing this, public interface `ConsumerFactory` is created to build consumer and pass into `consumeService`.

WIP as a follow up PR : Commit `Latency` Service

### CODE REVIEW

File: src/main/java/com/linkedin/kmf/KafkaMonitor.java [hunk 1]
```diff
@@ -79,8 +90,11 @@ public KafkaMonitor(Map<String, Map> testProps) throws Exception {
     List<MetricsReporter> reporters = new ArrayList<>();
     reporters.add(new JmxReporter(JMX_PREFIX));
     Metrics metrics = new Metrics(new MetricConfig(), reporters, new SystemTime());
-    metrics.addMetric(metrics.metricName("offline-runnable-count", METRIC_GROUP_NAME, "The number of Service/App that are not fully running"),
-      (config, now) -> _offlineRunnables.size());
+    Map<String, String> tags = new HashMap<>();
```
Comment (Lincong) [inline src/main/java/com/linkedin/kmf/KafkaMonitor.java]: Nit: If there is only one value in the map, you can use `Map<String, String> tags = new HashMap<>(1)`

File: src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java [hunk 2]
```diff
@@ -61,6 +64,7 @@ public SingleClusterMonitor(Map<String, Object> props, String name) throws Excep
   @Override
   public void start() {
     _topicManagementService.start();
+    _commitAvailabilityService.start();
     CompletableFuture<Void> completableFuture = _topicManagementService.topicManagementResult();
     completableFuture.thenRun(() -> {
```
Comment (Lincong) [inline src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java]: I know this line is not introduced as part of this PR. But I am just wondering why making the start of the produceService and consumerService async? Is the goal here to make the method `start()` return faster? I think so services their `start()` methods called synchronously whereas others don't. It feels a little strange.

Another problem with having async start is that what if `_produceService.start();` or `_consumeService.start();` throws exception(s)? No code is going to handle it and the whole `SingleClusterMonitor` seems to have started. But actually not (inconsistent state).

File: src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java [hunk 3]
```diff
@@ -85,6 +90,10 @@ public boolean isRunning() {
       isRunning = false;
       LOG.info("_topicManagementService not running.");
     }
+    if (!_commitAvailabilityService.isRunning()) {
```
Comment (Lincong) [inline src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java]: Why not return false when we encounter the first not-running service?

File: src/main/java/com/linkedin/kmf/KafkaMonitor.java (hunk 1 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/KafkaMonitor.java]: Updated.

File: src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java (hunk 2 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java]: Please reference the previous Pull Requests for justifications to asynchronous calls. They have to do with the fact that  `_topicManagementResult` has to `complete()` within `MultiClusterTopicManagementService` class before the `produceService` and `consumeService` can start otherwise the ordering of these calls get disoriented causing failures.;

File: src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java (hunk 3 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java]: This codestyle decision is actually a product of your suggestion back in PR number 187 (https://github.com/linkedin/kafka-monitor/pull/187)

Originally, the `isRunning` method returned false when the first service was not, in fact, running.

image

File: src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    }
+  }
+
+  private OffsetAndMetadata commitAndRetrieveOffsets(TopicPartition topicPartition, Map<TopicPartition, OffsetAndMetadata> offsetMap) throws Exception {
+    final AtomicBoolean callbackFired = new AtomicBoolean(false);
+    final AtomicReference<Exception> offsetCommitIssue = new AtomicReference<>(null);
+    OffsetAndMetadata committed = null;
+    long now = System.currentTimeMillis();
+    long deadline = now + TimeUnit.MINUTES.toMillis(1);
+    while (System.currentTimeMillis() < deadline) {
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java]: Why have a deadline?  Why not just have a thread that periodically commits offsets and then use the callback to measure commit time?

File: config/kafka-monitor.properties [hunk 5]
```diff
@@ -43,7 +43,7 @@
 {
   "single-cluster-monitor": {
     "class.name": "com.linkedin.kmf.apps.SingleClusterMonitor",
-    "topic": "kafka-monitor-topic",
+    "topic": "kafka-monitor-topic-name",
```
[the remark this answers is not in the document]
Comment (andrewchoi5) [reply config/kafka-monitor.properties]: reverted. was testing with different topic name locally.

File: src/main/java/com/linkedin/kmf/KafkaMonitor.java [hunk 6]
```diff
@@ -40,7 +42,7 @@
   private static final Logger LOG = LoggerFactory.getLogger(KafkaMonitor.class);
   public static final String CLASS_NAME_CONFIG = "class.name";
   private static final String METRIC_GROUP_NAME = "kafka-monitor";
-  private static final String JMX_PREFIX = "kmf";
+  private static final String JMX_PREFIX = "kmf.services";
```
[the remark this answers is not in the document]
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/KafkaMonitor.java]: this change was for the purposes of alignment of with other sensors metrics. will amend.

File: src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java [hunk 7]
```diff
@@ -99,6 +115,7 @@ public boolean isRunning() {
   @Override
   public void awaitShutdown() {
     _topicManagementService.awaitShutdown();
+    _commitAvailabilityService.awaitShutdown();
     _produceService.awaitShutdown();
     _consumeService.awaitShutdown();
   }
```
[the remark this answers is not in the document]
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java]: Done

File: src/main/java/com/linkedin/kmf/KafkaMonitor.java [hunk 8]
```diff
@@ -79,8 +90,11 @@ public KafkaMonitor(Map<String, Map> testProps) throws Exception {
     List<MetricsReporter> reporters = new ArrayList<>();
     reporters.add(new JmxReporter(JMX_PREFIX));
     Metrics metrics = new Metrics(new MetricConfig(), reporters, new SystemTime());
-    metrics.addMetric(metrics.metricName("offline-runnable-count", METRIC_GROUP_NAME, "The number of Service/App that are not fully running"),
-      (config, now) -> _offlineRunnables.size());
+    Map<String, String> tags = new HashMap<>(1);
+    tags.put("name", "kafka-monitor");
```
[the remark this answers is not in the document]
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/KafkaMonitor.java]: All other metrics and sensors have the following structure,
```kmf.services:name=single-cluster-monitor,type=produce-service:produce-availability-avg=1.0
kmf.services:name=single-cluster-monitor,type=consume-service:consume-availability-avg=1.0
kmf.services:name=single-cluster-monitor,type=commit-availability-service:offsets-committed-avg=1.0
kmf.services:name=single-cluster-monitor,type=commit-availability-service:failed-commit-offsets-avg=0.0
kmf.services:name=single-cluster-monitor,type=produce-service:records-produced-total=148.0
kmf.services:name=single-cluster-monitor,type=consume-service:records-consumed-total=116.0
kmf.services:name=single-cluster-monitor,type=consume-service:records-lost-total=0.0```

except for offline runnables:
kmf:name=kafka-monitor,type=kafka-monitor:offline-runnable-count=0.0

This fixes such inconsistency.

File: src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java (hunk 4 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java]: Revised to utilize period offset commits

File: src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.Assert;
+
+import static org.testng.Assert.assertNotNull;
+
+/**
+ * CommitAvailabilityService measures the availability of consume offset commits to the Kafka broker.
+ */
+public class CommitAvailabilityService implements Service {
```
[the remark this answers is not in the document]
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java]: Would we always want to run Commit Availability Service when KMF is running Consume Service? The expectations are that commit availabilities are needed on some occasions.

File: src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java [hunk 10]
```diff
@@ -175,11 +175,9 @@ private void initiateCommit() throws Exception {
   private OffsetAndMetadata commitAndRetrieveOffsets(TopicPartition topicPartition, Map<TopicPartition, OffsetAndMetadata> offsetMap) throws Exception {
     final AtomicBoolean callbackFired = new AtomicBoolean(false);
     final AtomicReference<Exception> offsetCommitIssue = new AtomicReference<>(null);
-    OffsetAndMetadata committed = null;
-    long now = System.currentTimeMillis();
-    long deadline = now + TimeUnit.MINUTES.toMillis(1);
-    while (System.currentTimeMillis() < deadline) {
-      // Call commitAsync, wait for a NON-NULL return value (see https://issues.apache.org/jira/browse/KAFKA-6183)
+    OffsetAndMetadata committed;
+    while (true) {
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java]: Should this check if the service has stopped?

File: src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+        commitAndRetrieveOffsets(topicPartition, _offsetsToCommit);
+
+      } catch (KafkaException kafkaException) {
+        LOG.error("Exception while trying to to async commit.", kafkaException);
+        _commitAvailabilityMetrics._failedCommitOffsets.record();
+      }
+    }
+  }
+
+  private OffsetAndMetadata commitAndRetrieveOffsets(TopicPartition topicPartition, Map<TopicPartition, OffsetAndMetadata> offsetMap) throws Exception {
```
[the remark this answers is not in the document]
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/CommitAvailabilityService.java]: Since it’s impossible to add testing for commitRetrieveOffsets without initializing a new CommitAvailabilityService instance in the Test  class. Majority of KMF code is badly structured (when originally written since its inception) for unit testing.

File: src/main/java/com/linkedin/kmf/services/CommitAvailabilityMetrics.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(CommitAvailabilityMetrics.class);
+
+  /**
+   * Metrics for Calculating the offset commit availability of a consumer.
+   * @param metrics the commit offset metrics
+   * @param tags the tags associated, i.e) kmf.services:name=single-cluster-monitor
+   */
+  CommitAvailabilityMetrics(final Metrics metrics, final Map<String, String> tags) {
+    LOG.info("{} called.", this.getClass().getSimpleName());
+    _offsetsCommitted = metrics.sensor("offsets-committed");
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/CommitAvailabilityMetrics.java]: I'm not sure how quantile metrics are computed with the metric library being used, but perhaps there should be  some kind of 99th percentile commit time.  Is this the CommitLatencyService?

File: src/main/java/com/linkedin/kmf/services/ConsumeMetrics.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+      DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(Collections.singleton(topicName));
+      Map<String, KafkaFuture<TopicDescription>> topicResultValues = describeTopicsResult.values();
+      KafkaFuture<TopicDescription> topicDescriptionKafkaFuture = topicResultValues.get(topicName);
+      TopicDescription topicDescription = null;
+      try {
+        topicDescription = topicDescriptionKafkaFuture.get();
+      } catch (InterruptedException | ExecutionException e) {
+        LOG.error("Exception occurred while retrieving the topic description.", e);
+      }
+      assert topicDescription != null;
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeMetrics.java]: Asserts are special instructions which are not normally executed by the JVM unless the -ea flag is set when the JVM starts up.  The default is is that the flag is not set.  So asserts in Java are really only useful if the check the assert is guarding is  something expensive like an O(n) check.  For O(1) checks just throwing an exception is almost always better.

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 14]
```diff
@@ -149,15 +147,52 @@ private void consume() throws Exception {
...
+        };
+
+        if (_offsetsToCommit != null) {
+          _consumer.commitAsync(_offsetsToCommit, commitCallback);
+        } else {
+          _consumer.commitAsync(commitCallback);
+        }
+        _commitAvailabilityMetrics._offsetsCommitted.record();
+
+        org.testng.Assert.assertNull(offsetCommitIssues.get(), "Offset commit failed.");
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: testng should not be used within actual production code.

If having a method that checks for a condition is more clear then perhaps the google guava library is the way to go.  https://github.com/google/guava/wiki/PreconditionsExplained  It's probably ok to add a dependency on that library KMF does not already have this dependency.

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 15]
```diff
@@ -110,7 +106,8 @@ public ConsumeService(Map<String, Object> props, String name, CompletableFuture<
     if (props.containsKey(ConsumeServiceConfig.CONSUMER_PROPS_CONFIG)) {
       props.forEach(consumerProps::putIfAbsent);
     }
-    _consumer = (KMBaseConsumer) Class.forName(consumerClassName).getConstructor(String.class, Properties.class).newInstance(topic, consumerProps);
+    _consumer = (KMBaseConsumer) Class.forName(consumerClassName).getConstructor(String.class, Properties.class).newInstance(_topic, consumerProps);
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Because the consume service is creating it's own consumer rather than using an injected consumer it's not possible to mock the consumer without lots of effort.  This is one of the challenges with unit testing this class.  If the consume service either takes a consumer or  consumer factory as a parameter then mocking becomes possible, which means unit testing becomes possible.

Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/ConsumeService.java]: In order to render unit testing possible for `ConsumeService`, I will make the `consumeService` take in consumer / consumer factory as an argument in a follow PR if that's feasible and okay with you.

File: src/main/java/com/linkedin/kmf/services/ConsumeMetrics.java (hunk 13 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/ConsumeMetrics.java]: Agreed. Revised.

File: src/main/java/com/linkedin/kmf/services/CommitAvailabilityMetrics.java (hunk 12 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/CommitAvailabilityMetrics.java]: The class in discussion is for the `CommitAvailabilityMetrics`. The `CommitLatencyService` will be sampling and determining the 99th percentile commit time for its calculation.

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java (hunk 14 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Amended to not use testng library within the prod codebase.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+ * Unit Testing for the Consume Service Class.
+ */
+public class ConsumeServiceTest {
+  private static final String BROKER_LIST = "localhost:9092";
+  private static final String ZK_CONNECT = "localhost:2181";
+  private static final String TOPIC = "kafka-monitor-topic-test";
+
+  @Test
+  public void lifecycleTest()
+      throws ClassNotFoundException, NoSuchMethodException, InstantiationException, IllegalAccessException,
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: Giant exception lists are hard to read and not very useful.  Also the test framework treats all exceptions the same unless it is told not to.  So it's safe to make this just throw Exception.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+             InvocationTargetException {
+
+    ConsumeService consumeService = consumeService();
+
+    /* Nothing should be started */
+    org.testng.Assert.assertFalse(consumeService.isRunning());
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: No one writes test code like org.testng.Assert; it's either Assert.assertFalse or do the static import of the methods used for assertions.

File: src/main/java/com/linkedin/kmf/services/ConsumerFactory.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.Map;
+import java.util.Properties;
+import java.util.Random;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+
+public class ConsumerFactory {
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumerFactory.java]: If ConsumerFactory is an interface then it will be a little easier to mock.  This implementation then moves to ConsumerFactoryImpl (or whatever you want to call it).

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+    /* Should be allowed to shutdown more than once. */
+    consumeService.awaitShutdown();
+    consumeService.awaitShutdown();
+    org.testng.Assert.assertFalse(consumeService.isRunning());
+
+  }
+
+
+  private ConsumeService consumeService()
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: Once the ConsumerFactory is an interface most of this code can go away and be replaced with code that returns a mocked consumer.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.testng.annotations.Test;
+
+
+/**
+ * Unit Testing for the Consume Service Class.
+ */
+public class ConsumeServiceTest {
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: Thanks for creating this test.  While this looks like it will test lifecycle will it test any of the commit availability code?

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 21]
```diff
@@ -149,15 +98,50 @@ private void consume() throws Exception {
...
+        final AtomicReference<Exception> offsetCommitIssues = new AtomicReference<>(null);
+        OffsetAndMetadata offsetAndMetadata;
+        /* Call commitAsync, wait for a NON-NULL return value (see https://issues.apache.org/jira/browse/KAFKA-6183) */
+        OffsetCommitCallback commitCallback = new OffsetCommitCallback() {
+          @Override
+          public void onComplete(Map<TopicPartition, OffsetAndMetadata> topicPartitionOffsetAndMetadataMap, Exception exception) {
+            if (exception != null) {
+              offsetCommitIssues.set(exception);
+            }
+            callbackFired.set(true);
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: When does this value get used?

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 22]
```diff
@@ -149,15 +98,50 @@ private void consume() throws Exception {
...
+          }
+        };
+
+        if (_offsetsToCommit != null && !_offsetsToCommit.isEmpty()) {
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Maybe don't commit offsets every message as it's kind of an extreme use case.  Perhaps select a time period, like every second or 5 seconds for  committing an offset.

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 23]
```diff
@@ -149,15 +98,50 @@ private void consume() throws Exception {
...
+          _baseConsumer.commitAsync(_offsetsToCommit, commitCallback);
+        } else {
+          _baseConsumer.commitAsync(commitCallback);
+        }
+        _commitAvailabilityMetrics._offsetsCommitted.record();
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Shouldn't this be in the offset commit Callback?

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 24]
```diff
@@ -149,15 +98,50 @@ private void consume() throws Exception {
...
+
+        offsetAndMetadata = _baseConsumer.committed(topicPartition);
+        if (offsetAndMetadata != null) {
+          _offsetsToCommit.putIfAbsent(topicPartition, offsetAndMetadata);
+        }
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Won't this commit the same offsets every time?

File: src/main/java/com/linkedin/kmf/KafkaMonitor.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/KafkaMonitor.java]: Revert.

File: src/main/java/com/linkedin/kmf/KafkaMonitor.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/KafkaMonitor.java]: sgtm.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java (hunk 20 above)
Comment (andrewchoi5) [reply src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: Revised.

File: build.gradle [hunk 25]
```diff
@@ -41,8 +41,10 @@ allprojects {
     compile 'net.savantly:graphite-client:1.1.0-RELEASE'
     compile 'com.timgroup:java-statsd-client:3.0.1'
     compile 'com.signalfx.public:signalfx-codahale:0.0.47'
+    compile 'org.mockito:mockito-core:2.24.0'
```
Comment (smccauliff) [inline build.gradle]: Why does this need a compile time dependency on mockito?  testCompile should be sufficient.

File: src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java [hunk 26]
```diff
@@ -245,7 +254,7 @@ private static ArgumentParser argParser() {
   public static void main(String[] args) throws Exception {
     ArgumentParser parser = argParser();
     if (args.length == 0) {
-      System.out.println(parser.formatHelp());
+      LOG.info(parser.formatHelp());
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/apps/SingleClusterMonitor.java]: This is command line help and needs to be reported to the user on standard output or standard error.

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 27]
```diff
@@ -7,157 +7,148 @@
...
+          }
+
+          /* Record the current time for the committed consumer offset */
+          _baseConsumer.updateLastCommit();
+        }
+
+        offsetAndMetadata = _baseConsumer.committed(topicPartition);
+        if (offsetAndMetadata != null) {
+          _offsetsToCommit.put(topicPartition, offsetAndMetadata);
+        }
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: I'm still confused about why this is needed and how the offsets for the consumer will not get stuck since it would always be committing the value it was committing before.

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 28]
```diff
@@ -7,157 +7,148 @@
...
+
+      /* Commit availability and commit latency service */
+      try {
+        TopicPartition topicPartition = new TopicPartition(_topic, partition);
+        final AtomicReference<Exception> offsetCommitIssues = new AtomicReference<>(null);
+        OffsetAndMetadata offsetAndMetadata;
+        /* Call commitAsync, wait for a NON-NULL return value (see https://issues.apache.org/jira/browse/KAFKA-6183) */
+        OffsetCommitCallback commitCallback = (topicPartitionOffsetAndMetadataMap, exception) -> {
+          if (exception != null) {
+            offsetCommitIssues.set(exception);
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Is offsetCommitIssues ever used?  Should the following be called?

```java
        LOG.error("Exception while trying to perform an asynchronous commit.", kafkaException);
        _commitAvailabilityMetrics._failedCommitOffsets.record();
```

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java [hunk 29]
```diff
@@ -7,157 +7,148 @@
...
+          }
+          _commitAvailabilityMetrics._offsetsCommitted.record();
+        };
+
+        long currTimeMillis = System.currentTimeMillis();
+
+        /* 5 seconds consumer offset commit interval. */
+        long timeDiffMillis = 5000;
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kmf/services/ConsumeService.java]: The more readable way of having time constants in the code is to do something like this:

```java
long timeDiffMillis = TimeUnit.SECONDS.toMillis(5);
```
@radai-rosenblatt  says there is another way using the Duration class that is more modern.

Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Amended.

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java (hunk 28 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Revised to record failedCommitOffsets

File: src/main/java/com/linkedin/kmf/services/ConsumeService.java (hunk 27 above)
Comment (andrewchoi5) [reply src/main/java/com/linkedin/kmf/services/ConsumeService.java]: Correct. Updated.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+
+    ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+    ScheduledFuture<?> scheduledFuture = executorService.schedule(new Runnable() {
+      @Override
+      public void run() {
+        Assert.assertNotNull(metrics.metrics().get(metrics.metricName("offsets-committed-total", METRIC_GROUP_NAME, tags)).metricValue());
+        Assert.assertNotEquals(metrics.metrics().get(metrics.metricName("offsets-committed-total", METRIC_GROUP_NAME, tags)).metricValue(), 0.0);
+      }
+    }, 4, TimeUnit.SECONDS);
+    Assert.assertNotNull(scheduledFuture);
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: This code is unlikely to be executed during the duration of the test since nothing waits for it to complete or checks that the exceptions that would be thrown from assertion failures are not there.  Was the intention something like this:
```java
   scheduledFuture.get();  // This will throw ExecutionException
```

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+
+  }
+
+  static final class FakeConsumerFactory implements ConsumerFactory {
+    final Map<String, Object> _props;
+
+    /* Required */
+    public FakeConsumerFactory(Map<String, Object> props) {
+      _props = props;
+    }
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: Why is this required?  If this is not required then this class can go away as it can just be mocked.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+
+    @Override
+    public AdminClient adminClient() {
+      return AdminClient.create(_props);
+    }
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: This can be mocked

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+
+    @Override
+    public int latencySlaMs() {
+      return 20000;
+    }
+
+    @Override
+    public KMBaseConsumer baseConsumer() {
+      return Mockito.mock(KMBaseConsumer.class);
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: How are expectations being added to this mocked consumer?  Shouldn't it return some messages or something.
This might be a little dated but it seems fairly comprehensive.
https://www.vogella.com/tutorials/Mockito/article.html

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java (hunk 30 above)
Comment (andrewchoi5) [reply src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: modified with `scheduledFuture.get();`

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java (hunk 33 above)
Comment (andrewchoi5) [reply src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: _Updated._

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java (hunk 31 above)
Comment (andrewchoi5) [reply src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: Done

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 34]
```diff
@@ -0,0 +1,186 @@
...
+      public String answer(InvocationOnMock invocation) {
+        testCommittedSuccessfully = true;
+        return "Mock Offset Committed Successfully.";
+      }
+    }).when(kmBaseConsumer).commitAsync(Mockito.any());
+
+    /* avro record to KmBaseConsumer record */
+    Mockito.when(kmBaseConsumer.receive()).thenReturn(
+        new BaseConsumerRecord(TOPIC, 2, 3,
+            Utils.jsonFromFields(TOPIC, 2, 3, "producerId", 3),
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: The first usage of "Utils.jsonFromFields" is not needed as this parameter is just the key for the message and not the message itself (i.e. the value).  The ProduceService just sends an integer encoded as a string as the key.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 35]
```diff
@@ -0,0 +1,186 @@
...
+    consumeService.start();
+    Assert.assertTrue(consumeService.isRunning());
+
+    long threadStartDelay = 2000;
+
+    /* Thread.sleep safe to do here instead of ScheduledExecutorService
+    *  We want to sleep current thread so that consumeService can start running for enough seconds. */
+    Thread.sleep(threadStartDelay);
+    Assert.assertNotNull(metrics.metrics().get(metrics.metricName("offsets-committed-total", METRIC_GROUP_NAME, tags)).metricValue());
+    Assert.assertTrue(testCommittedSuccessfully);
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: This doesn't test that the ConsumeService processed the message correctly and recorded a count.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java [hunk 36]
```diff
@@ -0,0 +1,186 @@
...
+    /* define return value */
+    long mockLastCommittedOffset = 0;
+    Mockito.when(kmBaseConsumer.lastCommitted()).thenReturn(mockLastCommittedOffset);
+    Mockito.doAnswer(new Answer() {
+      @Override
+      public String answer(InvocationOnMock invocation) {
+        testCommittedSuccessfully = true;
+        return "Mock Offset Committed Successfully.";
+      }
+    }).when(kmBaseConsumer).commitAsync(Mockito.any());
```
Comment (smccauliff) [inline src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: The callback supplied to commitAsync needs to be called inside this method.

This looks relevant.

https://www.baeldung.com/mockito-callbacks

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java (hunk 35 above)
Comment (andrewchoi5) [reply src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: Updated.

File: src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java (hunk 36 above)
Comment (andrewchoi5) [reply src/test/java/com/linkedin/kmf/services/ConsumeServiceTest.java]: Done.
